File: src/main/java/org/academy/api/client/gui/animation/Animator.java
package org.academy.api.client.gui.animation;

import java.util.ArrayList;
import java.util.List;

public abstract class Animator {
    long startTime = -1;
    long startDelay = 0;
    long duration = 300;
    boolean running = false;
    boolean paused = false;
    long pauseBeginTime = -1;
    private final List<AnimatorListener> listeners = new ArrayList<>();

    public void start() {
        if (running) return;
        running = true;
        paused = false;
        pauseBeginTime = -1;
        AnimationManager.startAnimation(this);
    }

    public void cancel() {
        if (!running) return;
        running = false;
        paused = false;
        AnimationManager.remove(this);
        var tempList = new ArrayList<>(listeners);
        for (var listener : tempList) listener.onAnimationCancel(this);
    }

    public void end() {
        if (!running) return;
        running = false;
        paused = false;
        AnimationManager.remove(this);
        var tempList = new ArrayList<>(listeners);
        for (var listener : tempList) listener.onAnimationEnd(this);
    }

    public void pause() {
        if (running && !paused) {
            paused = true;
            pauseBeginTime = -1;
        }
    }

    public void resume() {
        if (running && paused) paused = false;
    }

    public boolean isPaused() {
        return paused;
    }

    void onStartInternal() {
        running = true;
        paused = false;
        var tempList = new ArrayList<>(listeners);
        for (var listener : tempList) listener.onAnimationStart(this);
    }

    public boolean isRunning() {
        return running;
    }

    public void addListener(AnimatorListener listener) {
        listeners.add(listener);
    }

    public void removeListener(AnimatorListener listener) {
        listeners.remove(listener);
    }

    public Animator setDuration(long duration) {
        if (duration < 0) throw new IllegalArgumentException("Animators cannot have negative duration: " + duration);
        this.duration = duration;
        return this;
    }

    public long getDuration() {
        return duration;
    }

    public Animator setStartDelay(long startDelay) {
        if (startDelay < 0) throw new IllegalArgumentException("Animators cannot have negative start delay: " + startDelay);
        this.startDelay = startDelay;
        return this;
    }

    public long getStartDelay() {
        return startDelay;
    }

    abstract boolean tick(long currentTime);
}

File: src/main/kotlin/org/academy/api/client/gui/animation/Animator.kt
package org.academy.api.client.gui.animation

abstract class Animator {
    var startTime: Long = -1
    var startDelay: Long = 0
    var duration: Long = 300
    var isRunning: Boolean = false
    var isPaused: Boolean = false
    var pauseBeginTime: Long = -1
    private val listeners: MutableList<AnimatorListener> = ArrayList()

    open fun start() {
        if (this.isRunning) return
        this.isRunning = true
        this.isPaused = false
        pauseBeginTime = -1
        AnimationManager.startAnimation(this)
    }

    fun cancel() {
        if (!this.isRunning) return
        this.isRunning = false
        this.isPaused = false
        AnimationManager.remove(this)
        val tempList = ArrayList<AnimatorListener?>(listeners)
        for (listener in tempList) listener!!.onAnimationCancel(this)
    }

    fun end() {
        if (!this.isRunning) return
        this.isRunning = false
        this.isPaused = false
        AnimationManager.remove(this)
        val tempList = ArrayList<AnimatorListener?>(listeners)
        for (listener in tempList) listener!!.onAnimationEnd(this)
    }

    fun pause() {
        if (this.isRunning && !this.isPaused) {
            this.isPaused = true
            pauseBeginTime = -1
        }
    }

    fun resume() {
        if (this.isRunning && this.isPaused) this.isPaused = false
    }

    open fun onStartInternal() {
        this.isRunning = true
        this.isPaused = false
        val tempList = ArrayList<AnimatorListener?>(listeners)
        for (listener in tempList) listener!!.onAnimationStart(this)
    }

    fun addListener(listener: AnimatorListener) {
        listeners.add(listener)
    }

    fun removeListener(listener: AnimatorListener) {
        listeners.remove(listener)
    }

    open fun setDuration(duration: Long): Animator {
        require(duration >= 0) { "Animators cannot have negative duration: $duration" }
        this.duration = duration
        return this
    }

    open fun setStartDelay(startDelay: Long): Animator {
        require(startDelay >= 0) { "Animators cannot have negative start delay: $startDelay" }
        this.startDelay = startDelay
        return this
    }

    abstract fun tick(currentTime: Long): Boolean
}